src/server/shell/decoration/renderer.cpp

```diff
@@ -74,6 +80,300 @@ inline void render_row(
 }
 }
 
+class msd::Renderer::Text::Impl
+    : public Text
+{
+public:
+    Impl();
+    ~Impl();
+
+    void render(
+        Pixel* buf,
+        geom::Size buf_size,
+        std::string const& text,
+        geom::Point top_left,
+        geom::Height height_pixels,
+        Pixel color) override;
+
+private:
+    struct FontPath
+    {
+        std::string const filename;
+        std::vector<std::string> prefixes;
+    };
+
+    static std::vector<FontPath> const font_paths;
+    static std::vector<std::string> const font_path_search_paths;
+
+    std::mutex mutex;
+    FT_Library library;
+    FT_Face face;
+
+    void set_char_size(geom::Height height);
+    void rasterize_glyph(char32_t glyph);
+    void render_glyph(
+        Pixel* buf,
+        geom::Size buf_size,
+        FT_Bitmap const* glyph,
+        geom::Point top_left,
+        Pixel color);
+
+    static auto font_path() -> std::string;
+    static auto utf8_to_utf32(std::string const& text) -> std::u32string;
+};
+
+std::vector<msd::Renderer::Text::Impl::FontPath> const msd::Renderer::Text::Impl::font_paths{
+    FontPath{"Ubuntu-B.ttf", {
+        "ubuntu-font-family",   // Ubuntu < 18.04
+        "ubuntu",               // Ubuntu >= 18.04/Arch
+    }},
+    FontPath{"FreeSansBold.ttf", {
+        "freefont",             // Debian/Ubuntu
+        "gnu-free",             // Fedora/Arch
+    }},
+    FontPath{"DejaVuSans-Bold.ttf", {
+        "dejavu",               // Ubuntu (others?)
+        "",                     // Arch
+    }},
+};
+
+std::vector<std::string> const msd::Renderer::Text::Impl::font_path_search_paths{
+    "/usr/share/fonts/truetype",    // Ubuntu/Debian
+    "/usr/share/fonts/TTF",         // Arch
+    "/usr/share/fonts",             // Fedora/Arch
+};
```

Comment:
I know there are plans to make vector and string constexpr, but do we need to construct user-defined types here?

The code is simpler (no calls to `c_str()`) with a `char const* const ...font_path_search_paths[]`.